example/cpp/diplomat_runtime.hpp

```diff
@@ -83,17 +87,17 @@ class result {
     return std::holds_alternative<Err<E>>(this->val);
   };
 
-  std::optional<T> ok() const {
+  std::optional<T> ok() {
     if (!this->is_ok()) {
       return std::nullopt;
     }
-    return std::make_optional(std::get<Ok<T>>(this->val).inner);
+    return std::make_optional(std::move(std::get<Ok<T>>(std::move(this->val)).inner));
   };
-  std::optional<E> err() const {
+  std::optional<E> err() {
```

Comment:
If you are `std::move`ing a field out of the receiver (i.e. `this->val`), you should mark the method as taking an rvalue reference, `&&`, because otherwise the compiler may ignore the `std::move`, and then copy-construct instead of move-construct.  From the docs page for std::move:

> `std::move` is used to *indicate* that an object `t` may be "moved from", i.e. allowing the efficient transfer of resources from `t` to another object.

To do this, add `&&` after the method name but before the braces.

Reply:
If it helps, here's how I think of it:

- `foo() { ... }` in C++ = `&mut self` in Rust
- `foo() && { ... }` in C++ = `self` in Rust
- `foo() const { ... }` in C++ = `&self` in Rust

Reply:
huh, i could swear I'd fixed this part already. Oh well.

Reply:
Ah, unfortunately this forces folks to `std::move()` to call `.ok()` if they have an lvalue, but i guess that is to be expected

We could overload this and add a by-ref method as well.

Reply:
std::optional does similar overloading https://en.cppreference.com/w/cpp/utility/optional/value